### PR TITLE
hoon: comments, refactors $coin rendering, minor efficiency improvements

### PR DESCRIPTION
**Pull request by joemfb**
Spurred by a suggestion of @Fang-'s in #3576, I needed to reacquaint myself with the various helper cores in `|co`, the `$coin` renderer. I've found it hard to remember how those helpers work, so I went through them and added detailed comments.

Having read them closely, I noticed various inefficiencies (separate calls to `+div` and `+mod` instead of `+dvr`, a handrolled `+pow` implementation, &c). Those matter, as we run this code quite frequently. Once I made those changes I decided to modernize the surrounding syntax as well. Once I did that, I couldn't prevent myself from furthering refactoring parts of the implementation. Most of the changes are just mechanical reorganization, using new runes (like `=/` and `=?`).

Please review for style and comment clarity.

### CODE REVIEW

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:5967]: You missed some old syntax here, and the `=-` `^=` `^-` combo can become `=;`.

```hoon
          =;  gam=(pair tape tape)
            (weld p.gam ?:(=(0 q.p.lot) `tape`['0' ~] q.gam))
          ?+  hay  [~ ((ox-co [10 3] |=(a=@ ~(d ne a))) q.p.lot)]
            %b  [['0' 'b' ~] ((ox-co [2 4] |=(a=@ ~(d ne a))) q.p.lot)]
            %i  [['0' 'i' ~] ((d-co 1) q.p.lot)]
            %x  [['0' 'x' ~] ((ox-co [16 4] |=(a=@ ~(x ne a))) q.p.lot)]
            %v  [['0' 'v' ~] ((ox-co [32 5] |=(a=@ ~(x ne a))) q.p.lot)]
            %w  [['0' 'w' ~] ((ox-co [64 5] |=(a=@ ~(w ne a))) q.p.lot)]
```

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:5993]: Phrasing the arguments here as "zero-padding" is a bit confusing, since it's more like a _minimum length_, which we reach by zero-padding if necessary.

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:5994]: Probably worth noting it's '0x'-prefixed.

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:5989]: It's more like a "word list", right? The comment should probably hint at how this takes a `(list @)`, unlike everything else here.

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:6052]:
```suggestion
  ::  like +em-co, also producing number of digits
```

You might consider prefixing these comment sections with the relevant arm name, too, ie:

```suggestion
  ::  +ed-co: like +em-co, also producing number of digits
```

**Review by Fang- — COMMENTED** [review]
Wow, much better! Thanks for touching this up. Some suggestions below.

File: pkg/arvo/sys/hoon.hoon
Comment (ohAitch) [inline pkg/arvo/sys/hoon.hoon:5910]: Another `a/@` -> `a=@` block similar to the one @Fang- found

File: pkg/arvo/sys/hoon.hoon
Comment (joemfb) [reply pkg/arvo/sys/hoon.hoon]: I've made these comments more like our typical arm docs.

**Comment by joemfb** [comment]
Thanks, @Fang-, @ohAitch, I've updated with your feedback (in new commits for ease of review -- I'll squash later). PTAL.

**Comment by ohAitch** [comment]
Can confirm all the `a/` and `%b` and `{}` and `=-` look gone :P

I particularly like the `dvr` bit, hadn't realized destructuring `=/` made it in.

File: pkg/arvo/sys/hoon.hoon
Comment (Fang-) [inline pkg/arvo/sys/hoon.hoon:5989]: Being a little bit more verbose here probably wouldn't hurt. Something like:

```suggestion
      ::  +d-co: decimal, takes minimum output digits
```

**Review by Fang- — APPROVED** [review]
Changes look good! I'm picking a nit about a comment, but this should be fine to go in regardless.

**Comment by joemfb** [comment]
Good thing I waited for CI, looks like there are memory leaks in the `+dvr` jet.

**Comment by ohAitch** [comment]
Oh that's no good! Is the issue that [`u3k(u3i())`](https://github.com/urbit/urbit/blob/master/pkg/urbit/jets/c/dvr.c#L28) is redundant?

**Comment by joemfb** [comment]
Yeah, that's it. Same thing in `+sqt`, and in at least one place in the `|fl` jets.

**Comment by joemfb** [comment]
This PR is blocked on #3656, and should not be OTA'd until those changes are released and users have had a chance to upgrade.

File: pkg/arvo/sys/hoon.hoon
Comment (pilfer-pandex) [inline pkg/arvo/sys/hoon.hoon:5996]: thank you

File: pkg/arvo/sys/hoon.hoon
Comment (pilfer-pandex) [inline pkg/arvo/sys/hoon.hoon:6009]: TBH, I tend to find stuff approaching this density easier to read wide, but obviously I defer to your taste.

File: pkg/arvo/sys/hoon.hoon
Comment (pilfer-pandex) [inline pkg/arvo/sys/hoon.hoon:6007]: Interesting style.

**Review by pilfer-pandex — APPROVED** [review]

File: pkg/arvo/sys/hoon.hoon
Comment (joemfb) [reply pkg/arvo/sys/hoon.hoon:5928]: `+end`/`+rsh` loops are pleasant to write and easy to understand, but decidedly suboptimal since they involve a bunch of incremental allocations. The practical choices are to either loop with a counter variable and use `+cut` (as this code now does), or `+rip`/`+ripn` the atom into a list and then map or reduce as needed. The former is definitely more efficient for large atoms, the latter is usually nicer code, and maybe more efficient for small-to-midsize atoms. We could use some more fundamental atom operations to make this kind of code nicer while being efficient.

**Comment by joemfb** [comment]
This has been merged out-of-band, into `na-release/candidate`.

**Pull request merged**
